src/D2L.CodeStyle.Analyzers/Async/BlockingAnalyzer.cs

```diff
@@ -0,0 +1,359 @@
+			if( myMethodSymbol.IsAsync ) {
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.AsyncMethodCannotCallBlockingMethod,
+						invocation.GetLocation(),
+						invokedMethodSymbol.Name
+					)
+				);
+
+				// TODO: code fix that looks for an async equivalent and suggests calling it instead.
+				return;
+			}
+
+			ctx.ReportDiagnostic(
+				Diagnostic.Create(
+					Diagnostics.BlockingCallersMustBeBlocking,
+					invocation.GetLocation(),
+					invokedMethodSymbol.Name,
+					myMethodSymbol.Name
+				)
+			);
+
+			// TODO: code fix to add blocking
+		}
+
+		private static void WarnForUnusedBlockingAttributes(
+			CompilationAnalysisContext ctx,
+			INamedTypeSymbol blockingAttr,
+			ICollection<IMethodSymbol> havePossiblyUnusedBlockingAttribute,
+			ICollection<IMethodSymbol> callsSomethingBlocking
+		) {
+			// Emit a warning when [Blocking] looks unnecessary.
+
+			var unnecessaryThings = havePossiblyUnusedBlockingAttribute.Except( callsSomethingBlocking );
+
+			foreach( var thing in unnecessaryThings ) {
+				if( thing.IsAbstract || thing.IsAsync ) {
+					continue;
+				}
+
+				var attr = GetAttribute( thing, blockingAttr );
+
+				ctx.ReportDiagnostic(
+					Diagnostic.Create(
+						Diagnostics.UnnecessaryBlocking,
+						attr.ApplicationSyntaxReference.GetSyntax().GetLocation(),
+						thing.Name
+					)
+				);
+
+				// TODO: code fix to remove blocking
+			}
+		}
+
+		private static bool TryGetContainingMethod(
+			SyntaxNodeAnalysisContext ctx,
+			InvocationExpressionSyntax invocation,
+			IMethodSymbol invokedMethodSymbol,
+			out SyntaxNode decl
+		) {
+			// Find the thing that "owns" this invocation... hopefully its a
+			// method (maybe a local one) but it constructor or an initializer or...
+			// This is a bit shady.
+			var earliestAncestor = invocation.FirstAncestorOrSelf<SyntaxNode>(
+				d => d.Kind() switch {
+					SyntaxKind.LocalFunctionStatement => true,
+					SyntaxKind.SimpleLambdaExpression => true,
+					SyntaxKind.ParenthesizedLambdaExpression => true,
+					SyntaxKind.AnonymousMethodExpression => true,
+					_ => d is MemberDeclarationSyntax
+				}
+			);
+
+			if( earliestAncestor.IsKind( SyntaxKind.MethodDeclaration )
+			 || earliestAncestor.IsKind( SyntaxKind.LocalFunctionStatement )
+			) {
+				decl = earliestAncestor;
+				return true;
+			}
+
+			string place;
+
+			switch( earliestAncestor.Kind() ) {
+				case SyntaxKind.PropertyDeclaration:
+					// could be initializers or impls, this should be enough
+					// for the user to figure it out.
+					place = "properties";
+					break;
+				case SyntaxKind.FieldDeclaration:
+					place = "field initializers";
+					break;
+				case SyntaxKind.ConstructorDeclaration:
+					place = "constructors";
+					break;
+				case SyntaxKind.ParenthesizedLambdaExpression:
+				case SyntaxKind.SimpleLambdaExpression:
+					place = "lambdas";
+					break;
+				case SyntaxKind.AnonymousMethodExpression:
+					place = "delegates";
+					break;

```

Comment:
Wouldn't this be incompatible with say. `T ICache.Add<T>( Func<T> createAndReturn, ... )`, where `createAndReturn` is guaranteed to call something blocking? I this case, could alternatively call the blocking thing ahead of time and just capture the result in the lambda, but that doesn't work for say `GetSetAndCacheLocaly` where the fetch parameters are determined by the cache (filtering out keys that are already cached).

Reply:
I was thinking (but didn't communicate) that we'd ultimately need something like `StatelessFunc` for blocking funcs :(

Reply:
(which would still of course require changes here)